### PR TITLE
[Distributed] Ensure to swift_release during destroying decoded distributed call arguments

### LINKED ISSUE #65853: Remote target (for distributed actor) arguments are not properly deinitialized
Remote target arguments reconstructed on the remote side of the distributed actor are not properly deinitialised.
If the remote target argument is a struct with some class instance references then those class instances are never released, which cause a memory leak.

We implemented minimised [test](https://github.com/ordo-one/external-reproducers/tree/main/swift/distributed-system-leak)
It perform a remote call on the distributed actor with a struct argument having a reference to the ClassValue instance.
Can find the ClassValue instance initialized twice (first time on the caller side, second time on the remote side), but reinitialised only once (on the caller side):
```
git clone https://github.com/ordo-one/external-reproducers.git
Cloning into 'external-reproducers'...
remote: Enumerating objects: 135, done.
remote: Counting objects: 100% (135/135), done.
remote: Compressing objects: 100% (88/88), done.
remote: Total 135 (delta 49), reused 91 (delta 23), pack-reused 0
Receiving objects: 100% (135/135), 27.13 KiB | 524.00 KiB/s, done.
Resolving deltas: 100% (49/49), done.
aaa % cd external-reproducers/swift/distributed-system-leak 
distributed-system-leak % swift test
Building for debugging...
[5/5] Linking DistributedSystemPackageTests
Build complete! (7.65s)
Test Suite 'All tests' started at 2023-05-11 12:45:03.890
Test Suite 'DistributedSystemPackageTests.xctest' started at 2023-05-11 12:45:03.891
Test Suite 'DistributedSystemTests' started at 2023-05-11 12:45:03.891
Test Case '-[DistributedSystemTests.DistributedSystemTests test1]' started.
Test Case '-[DistributedSystemTests.DistributedSystemTests test1]' passed (0.002 seconds).
Test Suite 'DistributedSystemTests' passed at 2023-05-11 12:45:03.893.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.002) seconds
Test Suite 'DistributedSystemPackageTests.xctest' passed at 2023-05-11 12:45:03.893.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.002) seconds
Test Suite 'All tests' passed at 2023-05-11 12:45:03.893.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.003) seconds
assign<TestActor>: 1
actorReady<TestActor>: 1
resolve<TestActor>
init: 0x0000600002a301d0
recordArgument
remoteCallVoid
init: 0x0000600002a301f0
value=100
deinit: 0x0000600002a301d0
distributed-system-leak % 
```

We would expect the second ClassValue instance constructed on the remote side would be deinitialised, thus the output would also had a line
```
deinit: 0x0000600002a301f0
```

**Environment**
- Swift compiler version info
```
distributed-system-leak % swift --version
swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
Target: arm64-apple-macosx13.0
```
- Xcode version info 
- Deployment target:
```
distributed-system-leak % uname -a
Darwin mbpro.local 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000 arm64
```

### PR DESCRIPTION
**Pull request by ktoso**
Follow up to https://github.com/apple/swift/pull/65874
Resolves: https://github.com/apple/swift/issues/65853 now also ensuring deinits properly run.

Thanks for review @DougGregor 
Huge thanks to @xedin for helping debug the issue.

We must destroy before we deallocate, otherwise the destroy does nothing to refcount and we would not run deinitializers as expected.

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
Why?

**Comment by ktoso** [comment]
@DougGregor asked for it, I guess so it's the same as other things in this func, also in reverse? "Shouldn't matter _but_" I guess

**Comment by xedin** [comment]
The other thing is stack allocations which should be in reverse order, this one is just calls to destroy function that’s why I am asking the question here :)

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Yeah it really shouldn't matter for this one... I'll ask @DougGregor if he thinks we'd better do it like this anyway or nah.

**Comment by ktoso** [comment]
note to self; add some more tests to check various combinations of types

**Comment by ktoso** [comment]
I was adding more tests and actually see that deinits DON'T run and we have an unbalanced retain here still -- even if the memory is properly destroyed so we don't leak the object itself. We're in the Direct_Guaranteed case and take the param with `loadAsTake` so that's right... I don't know where else the bad retain could be from 🤔

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/IRGen/GenDistributed.cpp [hunk 1]
```diff
@@ -581,6 +581,13 @@ void DistributedAccessor::emitLoadOfWitnessTables(llvm::Value *witnessTables,
 }
 
 void DistributedAccessor::emitReturn(llvm::Value *errorValue) {
+  // Destroy loaded arguments.
+  // This MUST be done before deallocating, as otherwise we'd try to swift_release freed memory,
+  // which will be a no-op, however that also would mean we never drop retain counts to 0 and miss to run deinitializers of classes!
```
Comment (xedin) [inline lib/IRGen/GenDistributed.cpp:586]: Nit: Please clang-format this comment.

**Review by xedin — APPROVED** [review]

**Comment by ktoso** [comment]
Fixed formatting

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Pull request merged**